modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersionFreeList.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.storage.pagememory.mv;
+
+import static org.apache.ignite.internal.pagememory.PageIdAllocator.FLAG_AUX;
+import static org.apache.ignite.internal.pagememory.PageIdAllocator.INDEX_PARTITION;
+
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.ignite.internal.pagememory.PageMemory;
+import org.apache.ignite.internal.pagememory.evict.PageEvictionTracker;
+import org.apache.ignite.internal.pagememory.freelist.AbstractFreeList;
+import org.apache.ignite.internal.pagememory.metric.IoStatisticsHolder;
+import org.apache.ignite.internal.pagememory.util.PageLockListener;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+import org.apache.ignite.lang.IgniteLogger;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * {@link AbstractFreeList} for {@link RowVersion} instances.
+ */
+public class RowVersionFreeList extends AbstractFreeList<RowVersion> {
+    private static final IgniteLogger LOG = IgniteLogger.forClass(RowVersionFreeList.class);
+
+    private final IoStatisticsHolder statHolder;
+
+    /**
+     * Constructor.
+     *
+     * @param grpId              Group ID.
+     * @param pageMem            Page memory.
+     * @param lockLsnr           Page lock listener.
+     * @param metaPageId         Metadata page ID.
+     * @param initNew            {@code True} if new metadata should be initialized.
+     * @param pageListCacheLimit Page list cache limit.
+     * @param evictionTracker    Page eviction tracker.
+     * @param statHolder         Statistics holder to track IO operations.
+     * @throws IgniteInternalCheckedException If failed.
+     */
+    public RowVersionFreeList(
+            int grpId,
+            PageMemory pageMem,
+            PageLockListener lockLsnr,
+            long metaPageId,
+            boolean initNew,
+            @Nullable AtomicLong pageListCacheLimit,
+            PageEvictionTracker evictionTracker,
+            IoStatisticsHolder statHolder
+    ) throws IgniteInternalCheckedException {
+        super(
+                grpId,
+                "RowVersionFreeList_" + grpId,
+                pageMem,
+                null,
+                lockLsnr,
+                FLAG_AUX,
+                LOG,
+                metaPageId,
+                initNew,
+                pageListCacheLimit,
+                evictionTracker
+        );
+
+        this.statHolder = statHolder;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected long allocatePageNoReuse() throws IgniteInternalCheckedException {
+        return pageMem.allocatePage(grpId, INDEX_PARTITION, defaultPageFlag);
```

Comment:
I believe other structures store partitionId in a field, please check it. INDEX_PARTITION is only applicable in volatile case, persistent storage will need a valid partition number here.

Reply:
I copied it from `TableFreeList`, it uses the index partition. But in Ignite 2 the production code never uses index partition. I will talk to Kirill about this.

Reply:
Kirill told me the following: when a free list is one per region, we use index partition; if it's per partition, we use the data partition. In this case, it's global, so index partition seems fine.

If you have other concerns, please let me know.

Reply:
I now create these free lists one per MvStorage instance and each of them gets the corresponding partitionId.